### PR TITLE
Enable generic NamedTuples

### LINKED ISSUE #685: Support generic NamedTuples
I tried this for a generic NamedTuple:

```
S = TypeVar("S")
T = TypeVar("T")

class Pair(Generic[S, T],
    NamedTuple("Pair",[
        ('fst', S),
        ('second', T)])):
    def show(self):
        print(self.fst, self.second)

p = Pair("ab", 22)  # type: Pair[str, int]
```

Mypy raised an exception: `RuntimeError: TypeVarType is already analysed`.

Related question: is this the right way to define a generic NamedTuple?

Thanks!

### PR DESCRIPTION
**Pull request by ilevkivskyi**
Fixes #685

This builds on top of some infra I added for recursive types (Ref https://github.com/python/mypy/pull/13297). Implementation is based on the idea in https://github.com/python/mypy/pull/13297#issuecomment-1211317009. Generally it works well, but there are actually some problems for named tuples that are recursive. Special-casing them in `maptype.py` is a bit ugly, but I think this is best we can get at the moment.

An important note on runtime support for this: IIUC this should work in Python 3.11, and should work soon with `typing_extensions` on earlier versions. I didn't add any version checks here. If someone is interested in adding those, please do this.

### CODE REVIEW

**Comment by ilevkivskyi** [comment]
OK, since I added call syntax support for TypedDicts I also add support for NamedTuples as well. One can use
```python
NT = NamedTuple("NT", [("key", int), ("value", T)])
```
similarly to
```python
NT = Tuple[int, T]
```

There is however one problem: I use logic from what we do for proper classes. And it looks like this allows re-using bound type variables (e.g. in nested classes). Interestingly, it is prohibited for type aliases. So they will now be an exception, but also type alias behavior IMO matches PEP 484 specification on this.

Anyway, I will not change this now to not break things, but at some point we will need to unify things on type variable binding and reuse.

File: test-data/unit/check-namedtuple.test [hunk 1]
```diff
@@ -1162,3 +1162,147 @@ NT5 = NamedTuple(b'NT5', [('x', int), ('y', int)])  # E: "NamedTuple()" expects
...
+x: Tuple[int, ...]
+
+S = TypeVar("S")
+def foo(x: S, y: S) -> S: ...
+reveal_type(foo(nti, nti))  # N: Revealed type is "Tuple[builtins.int, builtins.int, fallback=__main__.NT[builtins.int]]"
+
+reveal_type(foo(nti, nts))  # N: Revealed type is "Tuple[builtins.int, builtins.object, fallback=__main__.NT[builtins.object]]"
+reveal_type(foo(nts, nti))  # N: Revealed type is "Tuple[builtins.int, builtins.object, fallback=__main__.NT[builtins.object]]"
+
+reveal_type(foo(nti, x))  # N: Revealed type is "builtins.tuple[builtins.int, ...]"
```
Comment (JelleZijlstra) [inline test-data/unit/check-namedtuple.test:1275]: The difference here (Tuple for fixed-length and builtins.tuple for variable-length) is unfortunate but that seems like a problem for a different PR.

**Review by JelleZijlstra — APPROVED** [review]

File: test-data/unit/check-namedtuple.test (hunk 1 above)
Comment (ilevkivskyi) [reply test-data/unit/check-namedtuple.test:1275]: IIRC in error messages we format them nicely, it is only `reveal_type()` that reflects the internal representation.

File: mypy/maptype.py [hunk 2]
```diff
@@ -1,8 +1,19 @@
 from typing import Dict, List
 
+import mypy.typeops
```
Comment (misuzu) [inline mypy/maptype.py:3]: This causes a circular import:
```
  File "venv/lib/python3.10/site-packages/mypy/report.py", line 20, in <module>
    from mypy import stats
  File "venv/lib/python3.10/site-packages/mypy/stats.py", line 13, in <module>
    from mypy.argmap import map_formals_to_actuals
  File "venv/lib/python3.10/site-packages/mypy/argmap.py", line 8, in <module>
    from mypy.maptype import map_instance_to_supertype
  File "venv/lib/python3.10/site-packages/mypy/maptype.py", line 3, in <module>
    import mypy.typeops
  File "venv/lib/python3.10/site-packages/mypy/typeops.py", line 15, in <module>
    from mypy.maptype import map_instance_to_supertype
ImportError: cannot import name 'map_instance_to_supertype' from partially initialized module 'mypy.maptype' (most likely due to a circular import) (venv/lib/python3.10/site-packages/mypy/maptype.py)
```

Comment (ilevkivskyi) [reply mypy/maptype.py:3]: Thanks for reporting! Do you have a (preferably simple) repro for this?

Comment (misuzu) [reply mypy/maptype.py:3]: Simply importing `mypy.report` without mypyc enabled should trigger this issue.

Comment (ilevkivskyi) [reply mypy/maptype.py:3]: See https://github.com/python/mypy/pull/13809 Btw just curious, why do you need to `import mypy.report` directly?

Comment (misuzu) [reply mypy/maptype.py:3]: It's one of [sanity checks in nixpkgs](https://github.com/NixOS/nixpkgs/blob/0490b307e5556a8804e710d0c744d29c80fbce48/pkgs/development/python-modules/mypy/default.nix#L66), to make sure mypy module is not broken.

**Pull request merged**